### PR TITLE
Support for TypeOf types

### PR DESCRIPTION
**Pull request by PappasBrent**
Adds support for Clang's `TypeOfExprType` and `TypeOfType` to VAST. This addresses issue #322.
- Adds the types `TypeOfExprType` and `TypeOfType`to `HighLevelTypes.td`. These types keep track of their desugared types (i.e., the type inside the call to `typeof()`), but do not yet track type qualifiers.
- Adds visit methods to the `CodeGenTypeVisitorMixin` for these new types.
- Add a single small test to VAST's test suite to verify that typeof types are lowered correctly.

### CODE REVIEW

File: include/vast/Dialect/HighLevel/HighLevelTypes.td [hunk 1]
```diff
@@ -442,4 +442,16 @@ def AdjustedType : HighLevelType< "Adjusted" > {
   let assemblyFormat = "`<` $original `,` $adjusted `>`";
 }
 
+def TypeOfExprType : HighLevelType< "TypeOfExpr" > {
+  let mnemonic = "typeofexpr";
```
Comment (pgoodman) [inline include/vast/Dialect/HighLevel/HighLevelTypes.td:446]: `typeof.expr`?

File: include/vast/Dialect/HighLevel/HighLevelTypes.td [hunk 2]
```diff
@@ -442,4 +442,16 @@ def AdjustedType : HighLevelType< "Adjusted" > {
...
+  let parameters = (ins "Type":$desugaredType);
+  let assemblyFormat = "`<` $desugaredType `>`";
+}
+
+def TypeOfType : HighLevelType< "TypeOf" > {
+  let mnemonic = "typeof";
```
Comment (pgoodman) [inline include/vast/Dialect/HighLevel/HighLevelTypes.td:452]: `typeof.type`

**Review by pgoodman — CHANGES_REQUESTED** [review]

**Comment by PappasBrent** [comment]
I've committed and pushed changes to this branch addressing @pgoodman's comments, but for some reason they are not yet appearing in this PR

File: include/vast/Translation/CodeGenTypeVisitor.hpp [hunk 3]
```diff
@@ -397,6 +443,14 @@ namespace vast::cg {
         auto VisitRValueReferenceType(const clang::RValueReferenceType *ty) -> mlir_type {
             return VisitRValueReferenceType(ty, ty->desugar().getQualifiers());
         }
+
+        auto VisitTypeOfExprType(const clang::TypeOfExprType *ty, qualifiers quals = {}) -> mlir_type {
```
[the remark this answers is not in the document]
Comment (Jezurko) [reply include/vast/Translation/CodeGenTypeVisitor.hpp:447]: Because we also need single argument version for clang AST TypeVisitor.

**Review by xlauko — APPROVED** [review]

**Pull request merged**